### PR TITLE
fix(itest): correct check of test run command exit status

### LINKED ISSUE #768: `repeated-integration-tests.sh` does not detect test failures
The script no longer correctly detects failed test runs.

### PR DESCRIPTION
**Pull request by andrewazores**
Fixes #768

### CODE REVIEW

File: repeated-integration-tests.sh [hunk 1]
```diff
@@ -28,15 +30,16 @@ while [ $runcount -lt $runs ]; do
...
     runcount=$((runcount+1))
+    podman logs cryostat-itest > "${server_logfile}"
 done
 
 echo
 echo "########################"
 echo "Test runs completed"
-echo "Failures: $failures"
-echo "Runs: $runs"
+echo "Failures: ${failures}"
```
Comment (jan-law) [inline repeated-integration-tests.sh:54]: I see `Failures: 0 ` - am I on the right branch? 
At least the log files aren't appearing in the untracked files.
```
jalaw@localhost ~/cryostat ((3cd5773f…))> git status
HEAD detached at aazores/repeated-itest-script-fixup
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.vscode/

nothing added to commit but untracked files present (use "git add" to track)
```
```
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< io.cryostat:cryostat >------------------------
[INFO] Building cryostat 2.1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (create-pod) @ cryostat ---
abaadac577085a9f3f75a5b5fabf22e953a11133120fda902e69690a4d0ef209
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (start-jfr-datasource) @ cryostat ---
b2f13434efa9a600a2f69984d70d1b413bbaab6d8dc6288e7569db7bd639297c
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (start-grafana) @ cryostat ---
b702fedc83b0644623d62ca222655191013dc01111efc8e0dc36dc0567bad93d
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (start-container) @ cryostat ---
92f2ca47dabd242e06a8f29c7af301531790a61254fca534973faea294abf3ac
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (wait-for-container) @ cryostat ---
curl: (56) Recv failure: Connection reset by peer
{"dashboardAvailable":true,"cryostatVersion":"v2.0.0-SNAPSHOT-236-gb9bace35-dirty","datasourceAvailable":true}[INFO] 
[INFO] --- maven-failsafe-plugin:2.22.2:integration-test (default-cli) @ cryostat ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------

...

[INFO] Running itest.NoopAuthV2IT
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.01 s <<< FAILURE! - in itest.NoopAuthV2IT
[ERROR] shouldRespond200  Time elapsed: 0.01 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: <{"meta":{"type":"application/json","status":"OKthisshouldfail"},"data":{"result":{"username":""}}}>
     but: was <{"meta":{"type":"application/json","status":"OK"},"data":{"result":{"username":""}}}>
	at itest.NoopAuthV2IT.shouldRespond200(NoopAuthV2IT.java:81)

[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   NoopAuthV2IT.shouldRespond200:81 
Expected: <{"meta":{"type":"application/json","status":"OKthisshouldfail"},"data":{"result":{"username":""}}}>
     but: was <{"meta":{"type":"application/json","status":"OK"},"data":{"result":{"username":""}}}>
[INFO] 
[ERROR] Tests run: 103, Failures: 1, Errors: 0, Skipped: 1
[INFO] 
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-jfr-datasource) @ cryostat ---
jfr-datasource-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-grafana) @ cryostat ---
grafana-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-container) @ cryostat ---
cryostat-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (destroy-pod) @ cryostat ---
d252daa6693a5234c1a3ef0d265fc8ddf4deec41d3f55c7dfbfa04f63fc4b3e0
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:53 min
[INFO] Finished at: 2021-12-02T13:25:28-05:00
[INFO] ------------------------------------------------------------------------
Error: no container with name or ID "cryostat-itest" found: no such container

########################
Test runs completed
Failures: 0
Runs: 1
```

Comment (andrewazores) [reply repeated-integration-tests.sh:54]: That's weird. It worked for me, although perhaps I had a test "Error" execution result and not a "Failure"? It seems weird that Maven's exit status would still be 0 on a Failure though. I'll see if I can reproduce this.

Comment (jan-law) [reply repeated-integration-tests.sh:54]: Hmm nevermind, a second `git fetch aazores` seemed to fix it.
```
jalaw@localhost ~/cryostat (close-rules-connections)> git fetch aazores
remote: Enumerating objects: 18, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 15 (delta 10), reused 13 (delta 8), pack-reused 0
Unpacking objects: 100% (15/15), 1.95 KiB | 29.00 KiB/s, done.
From github.com:andrewazores/cryostat
 * [new branch]        itest-container-health      -> aazores/itest-container-health
   3cd5773f..3d521e26  repeated-itest-script-fixup -> aazores/repeated-itest-script-fixup
jalaw@localhost ~/cryostat (close-rules-connections)> git checkout repeated-itest-script-fixup
M	src/test/java/itest/NoopAuthV2IT.java
Branch 'repeated-itest-script-fixup' set up to track remote branch 'repeated-itest-script-fixup' from 'aazores'.
Switched to a new branch 'repeated-itest-script-fixup'
```

Comment (andrewazores) [reply repeated-integration-tests.sh:54]: Huh.

```
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ReportIT.testGetReportShouldSendFile:160 » Execution io.vertx.ext.web.handler....
[INFO] 
[ERROR] Tests run: 103, Failures: 0, Errors: 1, Skipped: 1
[INFO] 
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-jfr-datasource) @ cryostat ---
[INFO] [stdout] jfr-datasource-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-grafana) @ cryostat ---
[INFO] [stdout] grafana-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-container) @ cryostat ---
[INFO] [stdout] cryostat-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (destroy-pod) @ cryostat ---
[INFO] [stdout] cad3544138cb18f9e36babd33f28f2fba919d15987bb8c9fe0faf645de0a0a65
[INFO] Segment walltime 243 s, segment projects service time 243 s, effective/maximum degree of concurrency 1.00/7
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:03 min (Wall Clock)
[INFO] Finished at: 2021-12-02T14:28:53-05:00
[INFO] ------------------------------------------------------------------------
Error: no container with name or ID "cryostat-itest" found: no such container

########################
Test runs completed
Failures: 0
Runs: 1
```

This did work earlier in the patch series. Not sure what I did to break it again.

File: repeated-integration-tests.sh [hunk 2]
```diff
@@ -15,22 +15,32 @@ function cleanup() {
...
-        exec:exec@start-jfr-datasource \
-        exec:exec@start-grafana \
-        exec:exec@start-container \
-        exec:exec@wait-for-container \
-        failsafe:integration-test \
-        exec:exec@stop-jfr-datasource \
-        exec:exec@stop-grafana \
-        exec:exec@stop-container \
-        exec:exec@destroy-pod |& tee -a "${client_logfile}"
+    mvn "${FLAGS[@]}" |& tee >($PIPECLEANER > ${client_logfile})
```
Comment (jan-law) [inline repeated-integration-tests.sh:43]: Does this error show up for you or is it just me?
```
repeated-integration-tests.sh: line 43: syntax error near unexpected token `('
repeated-integration-tests.sh: line 43: `    mvn "${FLAGS[@]}" |& tee >($PIPECLEANER > ${client_logfile})'
```

Comment (andrewazores) [reply repeated-integration-tests.sh:43]: I don't get that. That syntax is bash-specific - what are you using? The shebang line at the top of the file should cause it to be run with bash, but maybe not in this case.

Comment (jan-law) [reply repeated-integration-tests.sh:43]: Ah, running `bash repeated-integration-tests.sh` works fine, but `sh repeated-integration-tests.sh` results in the syntax error. Interesting

Comment (andrewazores) [reply repeated-integration-tests.sh:43]: Odd. What's `which sh` or `sh --version` say?

Comment (andrewazores) [reply repeated-integration-tests.sh:43]: Or most tellingly, `ls -l $(which sh)`

File: repeated-integration-tests.sh (hunk 1 above)
Comment (andrewazores) [reply repeated-integration-tests.sh]: Should be fixed now, I think.

File: repeated-integration-tests.sh (hunk 2 above)
Comment (jan-law) [reply repeated-integration-tests.sh]:
```
[jalaw@localhost cryostat]$ ls -l $(which sh)
lrwxrwxrwx. 1 root root 4 Jul 27  2020 /usr/bin/sh -> bash

[jalaw@localhost cryostat]$ sh --version
GNU bash, version 5.0.17(1)-release (x86_64-redhat-linux-gnu)
```

Comment (andrewazores) [reply repeated-integration-tests.sh]: Very weird.

```bash
~ $ ls -l $(which sh)
lrwxrwxrwx. 1 root root 4 Jan 25  2021 /usr/bin/sh -> bash

~ $ sh --version
GNU bash, version 5.1.0(1)-release (x86_64-redhat-linux-gnu)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

Comment (jan-law) [reply repeated-integration-tests.sh]: Here's the latest:
```
jalaw@localhost ~/cryostat (repeated-itest-script-fixup)> git pull
Updating 3d521e26..3b553584
Fast-forward
 repeated-integration-tests.sh => repeated-integration-tests.bash | 11 +++++++----
 repeated-unit-tests.bash                                         | 38 ++++++++++++++++++++++++++++++++++++++
 repeated-unit-tests.sh                                           | 26 --------------------------
 3 files changed, 45 insertions(+), 30 deletions(-)
 rename repeated-integration-tests.sh => repeated-integration-tests.bash (83%)
 create mode 100755 repeated-unit-tests.bash
 delete mode 100755 repeated-unit-tests.sh
jalaw@localhost ~/cryostat (repeated-itest-script-fixup)> sh repeated-integration-tests.bash 
repeated-integration-tests.bash: line 45: syntax error near unexpected token `('
repeated-integration-tests.bash: line 45: `    mvn "${FLAGS[@]}" |& tee >($PIPECLEANER > "${client_logfile}")'
jalaw@localhost ~/cryostat (repeated-itest-script-fixup) [1]> bash repeated-integration-tests.bash 
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< io.cryostat:cryostat >------------------------
[INFO] Building cryostat 2.1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (create-pod) @ cryostat ---
...
```

Comment (andrewazores) [reply repeated-integration-tests.sh]: And for me, all of `sh repeated-integration-tests.bash` (no tab complete on this one), `bash repeated-integration-tests.bash`, and `./repeated-integration-tests.bash` all seem to work fine.

`echo $SHELL`? `/bin/bash` for me.

Comment (andrewazores) [reply repeated-integration-tests.sh]: > Here's the latest:

Not really sure - I guess just start explicitly calling it with `bash` then. I could rewrite that part of the script to make it plain `sh`-compliant but there are several overall `bash`isms in there and it'd be a bit annoying to do.

File: repeated-integration-tests.sh (hunk 1 above)
Comment (jan-law) [reply repeated-integration-tests.sh]: Yes, all fixed!

**Review by jan-law — APPROVED** [review]

**Comment by andrewazores** [comment]
Sorry @jan-law . This wasn't as ready as I thought it was when I opened the PR. I think I'm finally done messing with it now.

File: repeated-integration-tests.bash [hunk 3]
```diff
@@ -40,14 +42,14 @@ DIR="$(dirname "$(readlink -f "$0")")"
 runcount=0
 while [ "${runcount}" -lt "${runs}" ]; do
     timestamp="$(date -Iminutes)"
-    client_logfile="$DIR/target/cryostat-itests-${timestamp}.client.log"
-    server_logfile="$DIR/target/cryostat-itests-${timestamp}.server.log"
+    client_logfile="$DIR/target/${POD_NAME}-${timestamp}.client.log"
```
Comment (jan-law) [inline repeated-integration-tests.bash:53]: I noticed when all the itests finish running, the test results are overwritten with the following. Where did the previous logs go?
`/home/jalaw/cryostat/target/cryostat-itests-2021-12-02T18:30-05:00.client.log`
```
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< io.cryostat:cryostat >------------------------
[INFO] Building cryostat 2.1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-jfr-datasource) @ cryostat ---
jfr-datasource-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-grafana) @ cryostat ---
grafana-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (stop-container) @ cryostat ---
cryostat-itest
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:exec (destroy-pod) @ cryostat ---
83c9173a91e43a40b7515c5f1532302d981a1455d17ddd3c9bc2b286ad5c73e0
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.403 s
[INFO] Finished at: 2021-12-02T18:34:03-05:00
[INFO] ------------------------------------------------------------------------
```

Comment (andrewazores) [reply repeated-integration-tests.bash:53]: Whoops. My "pipecleaner" change with the optionally preserved coloured output seems to have broken that. I think I fixed it.

**Review by jan-law — APPROVED** [review]
Everything's working now, looks good!

**Pull request merged**